### PR TITLE
Add command rewrite design doc

### PR DESCRIPTION
**Pull request by Oblarg**
The API details in the doc are somewhat outdated, but for purposes of justifying the rewrite it is still entirely accurate.

### CODE REVIEW

**Comment by Daltz333** [comment]
@AustinShalit

**Review by AustinShalit — DISMISSED** [review]

File: CommandRewriteDesignDoc.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+However, the libraries used in the framework are now more than a decade old.  Many of the particular choices in their implementation are obsolete, and others are messy, difficult-to-maintain, and/or confusingly-conceived.  Given the confusing nature of much of the implementation, bringing the API up to modern standards is a difficult task, and the libraries have ended up in a "stasis" of sorts.
+
+To document will describe the goals, design choices, and implementation of a ground-up rewrite of the Command-based libraries aimed at alleviating these issues.
+
+
+## Why rewrite the framework? (Problem Summary)
+
+The problems with the current command framework fall generally into three major categories:
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]: Off by one error?
```suggestion
The areas for improvement with the current command framework fall generally into four major categories:
```

File: CommandRewriteDesignDoc.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+1.  Poor readability/maintainability.
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]:
```suggestion
1.  Maintainability
```

File: CommandRewriteDesignDoc.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+2.  Poor encapsulation/separation of responsibilities.
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]:
```suggestion
2.  Encapsulation & separation of responsibilities
```

File: CommandRewriteDesignDoc.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+3.  Unnecessarily restrictive API design.
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]:
```suggestion
3.  Restrictive API design
```

File: CommandRewriteDesignDoc.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+4.  Clutter.
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md:18]:
```suggestion
4.  API Clutter
```

File: CommandRewriteDesignDoc.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+These three problems are interrelated, but I will summarize each in turn.
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]:
```suggestion
These four areas are interrelated, but I will summarize each in turn.
```

File: CommandRewriteDesignDoc.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+### Problem 4: The API Is Really, Really Cluttered
+
+Because of deeper issues with separation-of-concerns between the Command-based libraries and the `Sendable` implementation, many `Command`s have long lists of overloaded constructors and other cruft that make the code difficult to navigate and provide little useful functionality to the users.  The `PIDCommand` class is a particularly egregious example.
+
+Overall, the problems with the library run deep enough that an attempt to modify the existing structure would likely be futile.  Large, breaking changes are needed to address the concerns raised above.
+
+## Overall Design of the Rewrite
+
+The rewritten Command-based framework seeks to rectify the above problems.  A broad overview of the changes introduced in the rewrite is as follows:
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md:46]:
```suggestion
The rewritten Command-based framework seeks to improve on the areas listed above.  A broad overview of the changes introduced in the rewrite is as follows:
```

File: CommandRewriteDesignDoc.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+Design document
+
+The WPILib Command-based framework has been a staple of FRC Java and C++ robot code development for years, offering teams a relatively clean and sensible API for factoring complex robot code into simple, encapsulated chunks that interact in clear, comprehensible ways.  In addition to helping teams avoid the often substantial overhead of rolling their own state machine logic, it also serves as an important pedagogical tool for teaching teams about encapsulation, control flow, and design patterns.
+
+However, the libraries used in the framework are now more than a decade old.  Many of the particular choices in their implementation are obsolete, and others are messy, difficult-to-maintain, and/or confusingly-conceived.  Given the confusing nature of much of the implementation, bringing the API up to modern standards is a difficult task, and the libraries have ended up in a "stasis" of sorts.
+
+To document will describe the goals, design choices, and implementation of a ground-up rewrite of the Command-based libraries aimed at alleviating these issues.
+
+
+## Why rewrite the framework? (Problem Summary)
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md:11]: Can you please rewrite this section to be less condemning of the old framework? Teams have successfully used it for 10+ years and I think writing about it in this way is not good for it or WPILib. See some of my inline comments.

File: CommandRewriteDesignDoc.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+Trigger or(Trigger trigger);
+Trigger negate();
+```
+
+These allow `Trigger`s to be composed prior to button binding, a la:
+
+```java
+button1.and(button2).whenActive(...);
+```
+
```
Comment (AustinShalit) [inline CommandRewriteDesignDoc.md]: Extra EOL.
```suggestion
```

**Review by AustinShalit — CHANGES_REQUESTED** [review]

**Review by AustinShalit — APPROVED** [review]

**Pull request merged**